compiler/rustc_typeck/src/check/fn_ctxt/checks.rs

```diff
@@ -412,20 +346,603 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                }
+
+                if permutation_found {
+                    // Map unwrap to remove the first layer of Some
+                    let final_permutation: Vec<Option<usize>> =
+                        permutation.into_iter().map(|x| x.unwrap()).collect();
+                    return Some(Issue::Permutation(final_permutation));
+                }
+                return None;
+            };
+
+            // As we encounter issues, keep track of what we want to provide for the suggestion
+            let mut suggested_inputs: Vec<Option<String>> = vec![None; minimum_input_count];
+            let mut labels = vec![];
+            let mut suggestion_text = None;
+            let source_map = self.sess().source_map();
+
+            // Before we start looking for issues, eliminate any arguments that are already satisfied,
+            // so that an argument which is already spoken for by the input it's in doesn't
+            // spill over into another similarly typed input
+            // ex:
+            //   fn some_func(_a: i32, _a: i32) {}
+            //   some_func(1, "");
+            // Without this elimination, the first argument causes the second argument
+            // to show up as both a missing input and extra argument, rather than
+            // just an invalid type.
+            for (arg, inp) in
+                eliminate_satisfied(&mut compatibility_matrix, &mut input_indexes, &mut arg_indexes)
+            {
+                let arg_span = provided_args[arg].span;
+                let arg_text = source_map.span_to_snippet(arg_span).unwrap();
+                suggested_inputs[inp] = Some(arg_text);
+            }
+
+            // Until we've elimineated / satisfied all arguments/inputs
+            while input_indexes.len() > 0 || arg_indexes.len() > 0 {
+                // Check for the first relevant issue
+                match find_issue(&compatibility_matrix, &input_indexes, &arg_indexes) {
+                    Some(Issue::Invalid(idx)) => {
+                        // Eliminate the input and the arg, proposing a placeholder of the correct type
+                        let input_idx = input_indexes[idx];
+                        let expected_ty = expected_input_tys[input_idx];
+                        let input_ty = self.resolve_vars_if_possible(expected_ty);
+                        if input_ty.is_unit() {
+                            suggested_inputs[input_idx] = Some("()".to_string());
+                        } else {
+                            suggested_inputs[input_idx] = Some(format!("{{{}}}", input_ty));
+                        }
+                        if let Compatibility::Incompatible(coerce_error, subtyping_error) =
+                            &compatibility_matrix[idx][idx]
+                        {
+                            let provided_ty = final_arg_types[input_idx].map(|ty| ty.0).unwrap();
+                            let cause = &self.misc(provided_args[input_idx].span);
```

Comment:
We might want to create a specific `ObligationCauseCode` for this case, IIF there's value in that.